lib/ABI/FunctionType.cpp

```diff
@@ -740,28 +805,60 @@ Layout::Layout(const model::CABIFunctionType &Function) :
   })) {
 }
 
-Layout::Layout(const model::RawFunctionType &Function) {
-  // Lay register arguments out.
-  for (const model::NamedTypedRegister &Register : Function.Arguments)
-    Arguments.emplace_back().Registers = { Register.Location };
-
+static void
+fillLayoutFromRawFunctionTypeHelper(const model::RawFunctionType &Function,
+                                    abi::FunctionType::Layout &Output) {
   // Lay stack arguments out.
   if (Function.StackArgumentsType.isValid()) {
     const model::Type *OriginalStackType = Function.StackArgumentsType.get();
     auto *StackStruct = llvm::dyn_cast<model::StructType>(OriginalStackType);
     revng_assert(StackStruct,
                  "`RawFunctionType::StackArgumentsType` must be a struct.");
     typename Layout::Argument::StackSpan StackSpan{ 0, StackStruct->Size };
-    Arguments.emplace_back().Stack = std::move(StackSpan);
+    Output.Arguments.emplace_back().Stack = std::move(StackSpan);
   }
 
+  // Fill callee saved registers.
+  Output.CalleeSavedRegisters.resize(Function.PreservedRegisters.size());
+  llvm::copy(Function.PreservedRegisters, Output.CalleeSavedRegisters.begin());
+
+  // Set the final offset.
+  Output.FinalStackOffset = Function.FinalStackOffset;
+}
+
+Layout::Layout(const model::RawFunctionType &Function) {
+  // Lay register arguments out.
+  for (const model::NamedTypedRegister &Register : Function.Arguments)
+    Arguments.emplace_back().Registers = { Register.Location };
+
   // Lay the return value out.
   for (const model::TypedRegister &Register : Function.ReturnValues)
     ReturnValue.Registers.emplace_back(Register.Location);
 
-  // Fill callee saved registers.
-  CalleeSavedRegisters.resize(Function.PreservedRegisters.size());
-  llvm::copy(Function.PreservedRegisters, CalleeSavedRegisters.begin());
+  fillLayoutFromRawFunctionTypeHelper(Function, *this);
+}
+
+Layout::Layout(const model::RawFunctionType &Function, model::ABI::Values ABI) {
```

Comment:
Why do you need `ABI`? `RawFunctionType` is ABI-agnostic!

Reply:
This is one of overloads - it sorts argument registers in the "right" order based on the passed ABI.

Reply:
Mmmh, I understand how it could be useful, but I really want `RawFunctionType` (and all views on it) to be ABI-independent.
We're fine if it looks bad in the decompiled code, one more reason to make sure it never shows up!

Reply:
What would be the use case for that?
Consider that this way, we might have two distinct `llvm::Module`s containing functions with the same `RawFunctionType` but with a different `llvm::FunctionType` due to the fact that a different ABI was used to generate the layout. And this fact is not encoded anywhere in the model.

Reply:
The case I'm using it currently for is when I'm checking whether function semantics changed during a conversion (it also works as the `Layout` test). I just thought somebody else might find this useful as well.

If you're opposed against it that deeply, I'm Ok with moving it closer to the user.

Reply:
I suggest:
1. Consider if you can change tests not to rely on this behavior
2. If it's too much effort, `toRaw` must also add to the model the converted function. Then you can expose another function (used internally by `toRaw`) which has a name that suggest that you're not going to add it to the model *but* that other stuff is going to be added to the model nevertheless and put some good Doxygen.

Reply:
Ah, the problem is that you might get two different orderings?
Well, then the solution is that `Layout` should provide a predictable, ABI-independent ordering!
It can be trivial (register order) or you can define a custom one you like in `namespace Architecture { }` in `Model/Register.h`.

Reply:
The only difference is sorting - it's easy to do. So it's not the question of behaviour, it's the question of whether someone else might like the access to this functionality (I think, for example, @antoniofrighetto can find this useful at some point).

Reply:
Ah without changing the `Layout` itself? I thought you meant `orderArgumentRegisters` was to change the layout.

I honestly don't see a big use case for that, but if it the `Layout` of a `RawFunctionType` remains ABI-indepedent, I'm not against it.

Reply:
Yeah, initially I intended for it to change the layout. But if you're opposed extracting also works.

Reply:
What do you think?
https://github.com/revng/revng/pull/148/commits/4abfd3becf925c1f99a2dfed350ac7991781893d
https://github.com/revng/revng/pull/148/commits/ed4589cf2961ce2c3538d776ead4ff686f85e432

Reply:
Don't the belong to `model::ABI` at this point, more than `Layout`?
I haven't looked into testing yet, once I do, I will be able to comment on their usage there.

Reply:
Well, it does rely on ABI trait to sort them - so it "technically" has `abi` dependencies and `model` shouldn't depend on that.

Reply:
OK, sorry, I meant `abi`, but not `Layout`.

Reply:
Sure, I guess. It's just if we already have a free function like `SomeType doSomething(MyWellKnownObject, SomeFlag)`, it might as well just be `SomeType MyWellKnownObject::doSomething(SomeFlag)`. Don't you agree?

Reply:
Ah it's not a `static` method. You might drop the requirement for `Layout` and pass in a vector of registers.

Reply:
How about now?
https://github.com/revng/revng/pull/148/commits/a4b4a95015d243e55d79f82962238b0cff026404
https://github.com/revng/revng/pull/148/commits/980dd4d6ce8dc13814c1373924ff8357ef143211

Reply:
Works! Why not a `const Container &Registers` instead of `Container &&Registers`?

Reply:
It's a universal reference. Since container is template anyway, it'll deduce `const` if such a container is passed and so on. I guess there's no advantage of it now, since `SortedVector::BatchInserter::insert()` is pretty basic, but maybe one day this will allow avoiding a copy or something.

Reply:
What I mean is that using `std::forward` it's possible to propagate a r-value status down to the elements in the `for` loop - and then those elements can be moved when they are used. I didn't do that here though, since `insert` takes a `const T&` anyway.

Reply:
Uhm, OK, but `std::mov`ing `enum` values does not make much sense.
Universal references are not a widely known concept so I'd limit it to where it makes sense/infrastructure or very general code.

Reply:
Yeah, sure it's not that useful on an enum - but this function is not limited to those. It accepts a `sized_range`-like of any objects that are implicitly convertible to the said enum.

Reply:
I guess even just seeing something like
```cpp
template<typename T> void function(const T &);
```
feels so wrong to me because it's information lost for no reason. Even if that information was not used, it still doesn't feel good to throw it away.

Either way, here you go: https://github.com/revng/revng/pull/148/commits/0eefb2cc8ad17b662d51d26e282054c9ac917c9e

Reply:
I understand but, while you might be very familiar with universal references, their usage and implications, for the average dev (including me!) they are delicate and not super-straightforward and trigger some further thinking.

So, one of my key goals during review is: does this change has a good return on complexity coefficient?
That is, is this piece of code as simple as it could be without compromising on some aspect (functionality, performance)?
If the answer is no, I reject the extra complexity!